File: fore-jv/src/test/java/co/early/fore/core/ui/SyncTriggerTest.java
package co.early.fore.core.ui;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import static co.early.fore.core.ui.SyncTrigger.ResetRule.IMMEDIATELY;
import static co.early.fore.core.ui.SyncTrigger.ResetRule.NEVER;
import static co.early.fore.core.ui.SyncTrigger.ResetRule.ONLY_AFTER_REVERSION;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.mockito.Mockito.reset;


public class SyncTriggerTest {

    @Mock
    private Threshold mockThreshold;

    @Mock
    private Trigger mockTrigger;


    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);
    }



    @Test
    public void testCheckLazyReset_IMMEDIATELY() {

        SyncTrigger syncTrigger = new SyncTrigger(
                () -> mockTrigger.triggered(),
                () -> mockThreshold.shouldTrigger(),
                IMMEDIATELY);

        //check
        when(mockThreshold.shouldTrigger()).thenReturn(true);
        syncTrigger.check();
        verify(mockTrigger, times(1)).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);


        syncTrigger = new SyncTrigger(
                () -> mockTrigger.triggered(),
                () -> mockThreshold.shouldTrigger(),
                IMMEDIATELY);

        // checkLazy threshold breached, but trigger
        // is swallowed for first trigger
        when(mockThreshold.shouldTrigger()).thenReturn(true);
        syncTrigger.checkLazy();
        verify(mockTrigger, never()).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);

        //immediately reset and triggers
        when(mockThreshold.shouldTrigger()).thenReturn(true);
        syncTrigger.checkLazy();
        verify(mockTrigger, times(1)).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);

        //no trigger
        when(mockThreshold.shouldTrigger()).thenReturn(false);
        syncTrigger.checkLazy();
        verify(mockTrigger, never()).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);

        //immediately reset and triggers
        when(mockThreshold.shouldTrigger()).thenReturn(true);
        syncTrigger.checkLazy();
        verify(mockTrigger, times(1)).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);
    }

    @Test
    public void testCheckLazyReset_ONLY_AFTER_REVERSION() {

        SyncTrigger syncTrigger = new SyncTrigger(
                () -> mockTrigger.triggered(),
                () -> mockThreshold.shouldTrigger(),
                ONLY_AFTER_REVERSION);

        //check
        when(mockThreshold.shouldTrigger()).thenReturn(true);
        syncTrigger.check();
        verify(mockTrigger, times(1)).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);


        syncTrigger = new SyncTrigger(
                () -> mockTrigger.triggered(),
                () -> mockThreshold.shouldTrigger(),
                ONLY_AFTER_REVERSION);

        // checkLazy threshold breached, but trigger
        // is swallowed for first trigger
        when(mockThreshold.shouldTrigger()).thenReturn(true);
        syncTrigger.checkLazy();
        verify(mockTrigger, never()).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);

        //not yet reset, trigger not fired
        when(mockThreshold.shouldTrigger()).thenReturn(true);
        syncTrigger.checkLazy();
        verify(mockTrigger, never()).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);

        //reset, no trigger
        when(mockThreshold.shouldTrigger()).thenReturn(false);
        syncTrigger.checkLazy();
        verify(mockTrigger, never()).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);

        //now triggers once more
        when(mockThreshold.shouldTrigger()).thenReturn(true);
        syncTrigger.checkLazy();
        verify(mockTrigger, times(1)).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);
    }

    @Test
    public void testCheckLazyReset_NEVER() {

        SyncTrigger syncTrigger = new SyncTrigger(
                () -> mockTrigger.triggered(),
                () -> mockThreshold.shouldTrigger(),
                NEVER);

        //check
        when(mockThreshold.shouldTrigger()).thenReturn(true);
        syncTrigger.check();
        verify(mockTrigger, times(1)).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);


        syncTrigger = new SyncTrigger(
                () -> mockTrigger.triggered(),
                () -> mockThreshold.shouldTrigger(),
                NEVER);

        // checkLazy threshold breached, but trigger
        // is swallowed for first trigger
        when(mockThreshold.shouldTrigger()).thenReturn(true);
        syncTrigger.checkLazy();
        verify(mockTrigger, never()).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);

        //not reset, trigger not fired
        when(mockThreshold.shouldTrigger()).thenReturn(true);
        syncTrigger.checkLazy();
        verify(mockTrigger, never()).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);

        //threshold no longer breached, no trigger
        when(mockThreshold.shouldTrigger()).thenReturn(false);
        syncTrigger.checkLazy();
        verify(mockTrigger, never()).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);

        //never reset so no trigger still, even with threshold breached
        when(mockThreshold.shouldTrigger()).thenReturn(true);
        syncTrigger.checkLazy();
        verify(mockTrigger, never()).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);
    }

    @Test
    public void testCheckReset_IMMEDIATELY() {

        SyncTrigger syncTrigger = new SyncTrigger(
                () -> mockTrigger.triggered(),
                () -> mockThreshold.shouldTrigger(),
                IMMEDIATELY);

        //no trigger
        when(mockThreshold.shouldTrigger()).thenReturn(false);
        syncTrigger.check();
        verify(mockTrigger, never()).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);

        //trigger
        when(mockThreshold.shouldTrigger()).thenReturn(true);
        syncTrigger.check();
        verify(mockTrigger, times(1)).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);

        //no reset, still above threshold
        when(mockThreshold.shouldTrigger()).thenReturn(true);
        syncTrigger.check();
        verify(mockTrigger, times(1)).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);

        //reset, no trigger
        when(mockThreshold.shouldTrigger()).thenReturn(false);
        syncTrigger.check();
        verify(mockTrigger, never()).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);

        //trigger
        when(mockThreshold.shouldTrigger()).thenReturn(true);
        syncTrigger.check();
        verify(mockTrigger, times(1)).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);
    }

    @Test
    public void testCheckReset_ONLY_AFTER_REVERSION() {

        SyncTrigger syncTrigger = new SyncTrigger(
                () -> mockTrigger.triggered(),
                () -> mockThreshold.shouldTrigger(),
                ONLY_AFTER_REVERSION);

        //no trigger
        when(mockThreshold.shouldTrigger()).thenReturn(false);
        syncTrigger.check();
        verify(mockTrigger, never()).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);

        //trigger
        when(mockThreshold.shouldTrigger()).thenReturn(true);
        syncTrigger.check();
        verify(mockTrigger, times(1)).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);

        //no reset, still above threshold
        when(mockThreshold.shouldTrigger()).thenReturn(true);
        syncTrigger.check();
        verify(mockTrigger, never()).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);

        //reset, no trigger
        when(mockThreshold.shouldTrigger()).thenReturn(false);
        syncTrigger.check();
        verify(mockTrigger, never()).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);

        //trigger
        when(mockThreshold.shouldTrigger()).thenReturn(true);
        syncTrigger.check();
        verify(mockTrigger, times(1)).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);
    }

    @Test
    public void testCheckReset_NEVER() {

        SyncTrigger syncTrigger = new SyncTrigger(
                () -> mockTrigger.triggered(),
                () -> mockThreshold.shouldTrigger(),
                NEVER);

        //no trigger
        when(mockThreshold.shouldTrigger()).thenReturn(false);
        syncTrigger.check();
        verify(mockTrigger, never()).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);

        //trigger
        when(mockThreshold.shouldTrigger()).thenReturn(true);
        syncTrigger.check();
        verify(mockTrigger, times(1)).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);

        //no reset, still above threshold
        when(mockThreshold.shouldTrigger()).thenReturn(true);
        syncTrigger.check();
        verify(mockTrigger, never()).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);

        //below threshold, but never reset, no trigger
        when(mockThreshold.shouldTrigger()).thenReturn(false);
        syncTrigger.check();
        verify(mockTrigger, never()).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);

        //still no trigger
        when(mockThreshold.shouldTrigger()).thenReturn(true);
        syncTrigger.check();
        verify(mockTrigger, never()).triggered();
        verify(mockThreshold, times(1)).shouldTrigger();
        reset(mockTrigger);
        reset(mockThreshold);
    }


    class Threshold{
        public boolean shouldTrigger(){
            return true;
        }
    }

    class Trigger{
        public void triggered(){
            return;
        }
    }
}

File: fore-kt-android/src/test/java/co/early/fore/kt/core/ui/SyncTriggerTest.kt
package co.early.fore.kt.core.ui

import co.early.fore.kt.core.ui.synctrigger.ResetRule
import co.early.fore.kt.core.ui.synctrigger.SyncTrigger
import co.early.fore.kt.core.ui.synctrigger.SyncTriggerKeeper
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.verify
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Before
import org.junit.Test

class SyncTriggerTest {

    @MockK
    private lateinit var mockTriggeredWhen: () -> Boolean
    @MockK
    private lateinit var mockDoThis: () -> Unit

    @Before
    fun setup() = MockKAnnotations.init(this, relaxed = true)

    @Test
    fun `default SyncTrigger uses resetRule ONLY_AFTER_REVERSION`() {

        // arrange
        every { mockTriggeredWhen.invoke() } returns true
        val syncTrigger = SyncTrigger({ mockTriggeredWhen() }) {
            mockDoThis()
        }

        // act

        // assert
        assertEquals(ResetRule.ONLY_AFTER_REVERSION, syncTrigger.getResetRule())
    }

    @Test
    fun `on first successful trigger, check DOES fire`() {

        // arrange
        every { mockTriggeredWhen.invoke() } returns true
        val syncTrigger = SyncTrigger({ mockTriggeredWhen() }) {
            mockDoThis()
        }

        // act
        syncTrigger.check()

        // assert
        verify(exactly = 1) {
            mockDoThis()
        }
    }

    @Test
    fun `on first failed trigger, check DOES NOT fire`() {

        // arrange
        every { mockTriggeredWhen.invoke() } returns false
        val syncTrigger = SyncTrigger({ mockTriggeredWhen() }) {
            mockDoThis()
        }

        // act
        syncTrigger.check()

        // assert
        verify(exactly = 0) {
            mockDoThis()
        }
    }

    @Test
    fun `on first successful trigger, check lazy DOES NOT fire`() {

        // arrange
        every { mockTriggeredWhen.invoke() } returns true
        val syncTrigger = SyncTrigger({ mockTriggeredWhen() }) {
            mockDoThis()
        }

        // act
        syncTrigger.checkLazy()

        // assert
        verify(exactly = 0) {
            mockDoThis()
        }
    }

    @Test
    fun `on second successful trigger, check lazy DOES NOT fire`() {

        // arrange
        every { mockTriggeredWhen.invoke() } returns true
        val syncTrigger = SyncTrigger({ mockTriggeredWhen() }) {
            mockDoThis()
        }

        // act
        syncTrigger.checkLazy()
        syncTrigger.checkLazy()

        // assert
        verify(exactly = 0) {
            mockDoThis()
        }
    }

    @Test
    fun `on TRUE then FALSE trigger, check lazy DOES NOT fire`() {

        // arrange
        every { mockTriggeredWhen.invoke() } returns true andThen false
        val syncTrigger = SyncTrigger({ mockTriggeredWhen() }) {
            mockDoThis()
        }

        // act
        syncTrigger.checkLazy()
        syncTrigger.checkLazy()

        // assert
        verify(exactly = 0) {
            mockDoThis()
        }
    }

    @Test
    fun `on FALSE then TRUE trigger, check lazy DOES fire`() {

        // arrange
        every { mockTriggeredWhen.invoke() } returns false andThen true
        val syncTrigger = SyncTrigger({ mockTriggeredWhen() }) {
            mockDoThis()
        }

        // act
        syncTrigger.checkLazy()
        syncTrigger.checkLazy()

        // assert
        verify(exactly = 1) {
            mockDoThis()
        }
    }

    @Test
    fun `on TRUE, TRUE, TRUE triggers, check and reset ONLY_AFTER_REVERSION fires only once`() {

        // arrange
        every { mockTriggeredWhen.invoke() } returns true
        val syncTrigger = SyncTrigger({ mockTriggeredWhen() }) {
            mockDoThis()
        }.resetRule(ResetRule.ONLY_AFTER_REVERSION)

        // act
        syncTrigger.check()
        syncTrigger.check()
        syncTrigger.check()

        // assert
        verify(exactly = 1) {
            mockDoThis()
        }
    }

    @Test
    fun `on TRUE, TRUE, TRUE triggers, check lazy and reset ONLY_AFTER_REVERSION does NOT fire`() {

        // arrange
        every { mockTriggeredWhen.invoke() } returns true
        val syncTrigger = SyncTrigger({ mockTriggeredWhen() }) {
            mockDoThis()
        }.resetRule(ResetRule.ONLY_AFTER_REVERSION)

        // act
        syncTrigger.checkLazy()
        syncTrigger.checkLazy()
        syncTrigger.checkLazy()

        // assert
        verify(exactly = 0) {
            mockDoThis()
        }
    }

    @Test
    fun `on TRUE, FALSE, TRUE triggers, check and reset ONLY_AFTER_REVERSION fires twice`() {

        // arrange
        every { mockTriggeredWhen.invoke() } returns true andThen false andThen true
        val syncTrigger = SyncTrigger({ mockTriggeredWhen() }) {
            mockDoThis()
        }.resetRule(ResetRule.ONLY_AFTER_REVERSION)

        // act
        syncTrigger.check()
        syncTrigger.check()
        syncTrigger.check()

        // assert
        verify(exactly = 2) {
            mockDoThis()
        }
    }

    @Test
    fun `on TRUE, FALSE, TRUE triggers, check lazy and reset ONLY_AFTER_REVERSION fires once`() {

        // arrange
        every { mockTriggeredWhen.invoke() } returns true andThen false andThen true
        val syncTrigger = SyncTrigger({ mockTriggeredWhen() }) {
            mockDoThis()
        }.resetRule(ResetRule.ONLY_AFTER_REVERSION)

        // act
        syncTrigger.checkLazy()
        syncTrigger.checkLazy()
        syncTrigger.checkLazy()

        // assert
        verify(exactly = 1) {
            mockDoThis()
        }
    }

    @Test
    fun `on TRUE, TRUE, TRUE triggers, check and reset IMMEDIATELY fires each time`() {

        // arrange
        every { mockTriggeredWhen.invoke() } returns true
        val syncTrigger = SyncTrigger({ mockTriggeredWhen() }) {
            mockDoThis()
        }.resetRule(ResetRule.IMMEDIATELY)

        // act
        syncTrigger.check()
        syncTrigger.check()
        syncTrigger.check()

        // assert
        verify(exactly = 3) {
            mockDoThis()
        }
    }

    @Test
    fun `on TRUE, TRUE, TRUE triggers, check lazy and reset IMMEDIATELY fires each time except first`() {

        // arrange
        every { mockTriggeredWhen.invoke() } returns true
        val syncTrigger = SyncTrigger({ mockTriggeredWhen() }) {
            mockDoThis()
        }.resetRule(ResetRule.IMMEDIATELY)

        // act
        syncTrigger.checkLazy()
        syncTrigger.checkLazy()
        syncTrigger.checkLazy()

        // assert
        verify(exactly = 2) {
            mockDoThis()
        }
    }

    @Test
    fun `on TRUE, FALSE, TRUE triggers, check and reset IMMEDIATELY fires twice`() {

        // arrange
        every { mockTriggeredWhen.invoke() } returns true andThen false andThen true
        val syncTrigger = SyncTrigger({ mockTriggeredWhen() }) {
            mockDoThis()
        }.resetRule(ResetRule.IMMEDIATELY)

        // act
        syncTrigger.check()
        syncTrigger.check()
        syncTrigger.check()

        // assert
        verify(exactly = 2) {
            mockDoThis()
        }
    }

    @Test
    fun `on TRUE, FALSE, TRUE triggers, check lazy and reset IMMEDIATELY fires once`() {

        // arrange
        every { mockTriggeredWhen.invoke() } returns true andThen false andThen true
        val syncTrigger = SyncTrigger({ mockTriggeredWhen() }) {
            mockDoThis()
        }.resetRule(ResetRule.IMMEDIATELY)

        // act
        syncTrigger.checkLazy()
        syncTrigger.checkLazy()
        syncTrigger.checkLazy()

        // assert
        verify(exactly = 1) {
            mockDoThis()
        }
    }

    @Test
    fun `on TRUE, TRUE, TRUE triggers, check and reset NEVER, fires one time only`() {

        // arrange
        every { mockTriggeredWhen.invoke() } returns true
        val syncTrigger = SyncTrigger({ mockTriggeredWhen() }) {
            mockDoThis()
        }.resetRule(ResetRule.NEVER)

        // act
        syncTrigger.check()
        syncTrigger.check()
        syncTrigger.check()

        // assert
        verify(exactly = 1) {
            mockDoThis()
        }
    }

    @Test
    fun `on TRUE, TRUE, TRUE triggers, check lazy and reset NEVER, does NOT fire`() {

        // arrange
        every { mockTriggeredWhen.invoke() } returns true
        val syncTrigger = SyncTrigger({ mockTriggeredWhen() }) {
            mockDoThis()
        }.resetRule(ResetRule.NEVER)

        // act
        syncTrigger.checkLazy()
        syncTrigger.checkLazy()
        syncTrigger.checkLazy()

        // assert
        verify(exactly = 0) {
            mockDoThis()
        }
    }

    @Test
    fun `on TRUE, FALSE, TRUE triggers, check and reset NEVER, fires one time only`() {

        // arrange
        every { mockTriggeredWhen.invoke() } returns true andThen false andThen true
        val syncTrigger = SyncTrigger({ mockTriggeredWhen() }) {
            mockDoThis()
        }.resetRule(ResetRule.NEVER)

        // act
        syncTrigger.check()
        syncTrigger.check()
        syncTrigger.check()

        // assert
        verify(exactly = 1) {
            mockDoThis()
        }
    }

    @Test
    fun `on TRUE, FALSE, TRUE triggers, check lazy and reset NEVER, does NOT fire`() {

        // arrange
        every { mockTriggeredWhen.invoke() } returns true andThen false andThen true
        val syncTrigger = SyncTrigger({ mockTriggeredWhen() }) {
            mockDoThis()
        }.resetRule(ResetRule.NEVER)

        // act
        syncTrigger.checkLazy()
        syncTrigger.checkLazy()
        syncTrigger.checkLazy()

        // assert
        verify(exactly = 0) {
            mockDoThis()
        }
    }

    @Test
    fun `on trigger with keep, first kept value is null`() {

        // arrange
        var keptValue: Int? = null
        val syncTrigger = SyncTriggerKeeper<Int>(
            { keeper ->
                keeper.swap { previous ->
                    keptValue = previous
                    99
                }
                true
            })
        { mockDoThis() }.resetRule(ResetRule.NEVER)

        // act
        syncTrigger.check()

        // assert
        assertNull(keptValue)
    }

    @Test
    fun `on trigger with keep, second kept value is correct`() {

        // arrange
        var keptValue: Int? = null
        val syncTrigger = SyncTriggerKeeper<Int>(
            { keeper ->
                keeper.swap { previous ->
                    keptValue = previous
                    73
                }
                true
            })
        { mockDoThis() }.resetRule(ResetRule.NEVER)

        // act
        syncTrigger.check()
        syncTrigger.check()

        // assert
        assertEquals(73, keptValue)
    }

    @Test
    fun `on 3 triggers with keeps, triggeredWhen receives correct value`() {

        // arrange
        var keptValue = -1
        val syncTrigger = SyncTriggerKeeper<Int>(
            { keeper ->
                keeper.swap { previous ->
                    keptValue = previous ?: 0
                    val newValueToKeep = keptValue + 10
                    newValueToKeep
                }
                true
            })
        { mockDoThis() }.resetRule(ResetRule.NEVER)

        // act
        syncTrigger.check()
        syncTrigger.check()
        syncTrigger.check()
        syncTrigger.check()

        // assert
        assertEquals(30, keptValue)
    }

    @Test
    fun `on trigger with keep, first swap indicates change`() {

        // arrange
        var change = false
        val syncTrigger = SyncTriggerKeeper<Int>(
            { keeper ->
                change = keeper.swap {
                    val newValue = 73
                    newValue
                }
                true
            })
        { mockDoThis() }.resetRule(ResetRule.NEVER)

        // act
        syncTrigger.check()

        // assert
        assertEquals(true, change)
    }

    @Test
    fun `on trigger with changing kept value, swap indicates change`() {

        // arrange
        var change = false
        var keptValue: Int
        val syncTrigger = SyncTriggerKeeper<Int>(
            { keeper ->
                change = keeper.swap { previous ->
                    keptValue = previous ?: 0
                    val newValueToKeep = keptValue + 10
                    newValueToKeep
                }
                true
            })
        { mockDoThis() }.resetRule(ResetRule.NEVER)

        // act
        syncTrigger.check()
        syncTrigger.check()

        // assert
        assertEquals(true, change)
    }

    @Test
    fun `on trigger with same kept value, swap indicates no change`() {

        // arrange
        var change = false
        val syncTrigger = SyncTriggerKeeper<Int>(
            { keeper ->
                change = keeper.swap {
                    7
                }
                true
            })
        { mockDoThis() }.resetRule(ResetRule.NEVER)

        // act
        syncTrigger.check()
        syncTrigger.check()

        // assert
        assertEquals(false, change)
    }
}
